### PR TITLE
Fast Routes Command

### PR DESCRIPTION
**Pull request by elorest**
## Description
* Parses routes file and generates currect output.
* Removes dependency on amber core and doesn't have to compile anything so it's fast.
* Grabs the pipeline for each block
* Parses all basic verb routes [get, post, put, patch, etc]
* Parses resources and creates all verbs and actions.
* Parses `except` and `only` and doesn't add them to the list.
* Still uses @eliasjpr's routes command and prints out with the sexy table he wrote.

## To Test
```sh
git checkout is/fast-routes-command
make
cd ..
amber new fast_routes
cd fast_routes
amber g scaffold Cat name:string age:integer
amber g scaffold Dog name:string age:integer
amber routes
```
Go add `except: [:index, :show]` or `only: [:index, :show, :new, :create]` to config/routes.cr` to test those.

### Example:
```cr
  routes :web do
    resources "/birds", BirdController
    resources "/dogs", DogController, except: [:create, :update, :destroy]
    resources "/cats", CatController, only: [:index, :show]
    get "/", HomeController, :index
  end
```
Enjoy the speed!

2017-06-10 at 1 11 am

### CODE REVIEW

**Comment by elorest** [comment]
2017-06-10 at 1 11 am

File: src/amber_cmd/commands/routes.cr [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        puts "Error: Not valid project root directory.".colorize(:red)
+        puts "Run `amber routes` in project root directory.".colorize(:light_blue)
+        puts "Good bye :("
+        exit 1
+      end
+
+      private def parse_routes
+        File.read_lines("config/routes.cr").each do |line|
+          case line.strip
+          when .starts_with?("routes")
```
Comment (elorest) [inline src/amber_cmd/commands/routes.cr]: Yes.

File: src/amber_cmd/commands/routes.cr [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        print_routes_table
+      rescue
+        puts "Error: Not valid project root directory.".colorize(:red)
+        puts "Run `amber routes` in project root directory.".colorize(:light_blue)
+        puts "Good bye :("
+        exit 1
+      end
+
+      private def parse_routes
+        File.read_lines("config/routes.cr").each do |line|
```
Comment (elorest) [inline src/amber_cmd/commands/routes.cr]: It wouldn't work in that case but the other implementation will fail as well if files aren't where they're expected or one of many things changes in the router code.

File: src/amber_cmd/commands/routes.cr [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+          case line.strip
+          when .starts_with?("routes")
+            set_pipe(line)
+          when .starts_with?("get")
```
Comment (elorest) [inline src/amber_cmd/commands/routes.cr]: Hmmm. There's a lot of other lines in the routes.cr though and the regex is just looking for patterns not specific words. It's possible that something else could be parsed as a route unless we specifically pick select it. We could always through this logic into set_route though.

File: src/amber_cmd/commands/routes.cr [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+          when .starts_with?("delete")
+            set_route(line)
+          when .starts_with?("resources")
+            set_resources(line)
+          end
+        end
+      end
+
+      private def set_route(l)
+        if md = l.to_s.match(/(\w+)\s+\"([^\"]+)\",\s*(\w+),\s*:(\w+)/)
```
Comment (elorest) [inline src/amber_cmd/commands/routes.cr]: Yea. It doesn't currently take into account parens though so get("/cat", CatController,... Will fail currently. Easy fix though.

File: src/amber_cmd/commands/routes.cr [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+          route = {"Verb" => md[1]?.to_s}
+          route["Controller"] = md[3]?.to_s
+          route["Action"] = md[4]?.to_s
+          route["Pipeline"] = last_pipe.to_s
+          route["Scope"] = ""
```
Comment (elorest) [inline src/amber_cmd/commands/routes.cr]: I couldn't find any examples of scopes outside of pipelines to write this against.

File: src/amber_cmd/commands/routes.cr [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          route["Resource"] = md[2]?.to_s
+          routes << route
+        end
+      end
+
+      private def set_resources(l)
+        if md = l.to_s.match(/(\w+)\s+\"([^\"]+)\",\s*(\w+)(?:,\s*(\w+)\:\s*\[([^\]]+)\])?/)
```
Comment (elorest) [inline src/amber_cmd/commands/routes.cr]: Same. That will work fine.

**Comment by elorest** [comment]
If we change anything drastic about how the router works both methods will require us to make changes to the cmd side. As to pre optimizing 3 seconds is a long time even for a rails command. This way is far simpler requires no hacks and is fast. Scoping can be added as soon as I figure out what that looks like. If you'd like we could give it a list of possible routes files to parse although I do think we should be opinionated about out file structure.

**Comment by elorest** [comment]
Ok I just figured out what scopes look like.

```cr
  routes :static, "/page" do
    # Each route is defined as follow
    # verb resource : String, controller : Symbol, action : Symbol
    get "/*", StaticController, :index
  end
```

File: src/amber_cmd/commands/routes.cr (hunk 2 above)
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: The first approach works regardless from where the routes is defined since is not looking at `./config/routes.cr`

**Comment by eliasjpr** [comment]
The first implementation would not required us to change the command because the routes passed to the command from the app and also would work no matter where you define the routes it doesn't have knowledge of the `./config/routes`

File: config/application.cr [hunk 7]
```diff
@@ -0,0 +1,20 @@
+Amber::Server.instance.config do |app|
```
Comment (drujensen) [inline config/application.cr]: Is this file needed?

File: config/database.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+pg:
```
Comment (drujensen) [inline config/database.yml]: can we remove this?

File: config/routes.cr [hunk 9]
```diff
@@ -0,0 +1,33 @@
+Amber::Server.instance.config do |app|
```
Comment (drujensen) [inline config/routes.cr]: And this?

File: src/amber_cmd/commands/routes.cr [hunk 10]
```diff
@@ -23,30 +35,79 @@ module Amber::CMD
...
-          AMBER_CMD_PROCESS_ROUTES = true
-          puts Amber::Server.routes.to_json
-        CODE
-        routes_json = `crystal eval #{code.gsub("\n", ";").inspect}`
-        return JSON.parse(routes_json) unless routes_json.empty?
-        puts "No routes defined!"
+      private def set_route(l)
+        if md = l.to_s.match(VERB_ROUTE_REGEX)
+          return unless ACTION_MAPPING.keys.includes?(md[1]?.to_s)
+          route = {"Verb" => md[1]?.to_s}
```
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: Can we subtract the duplicate code to a method? This is in 2 places

File: src/amber_cmd/commands/routes.cr [hunk 11]
```diff
@@ -23,30 +35,79 @@ module Amber::CMD
...
+        end
+      end
 
-          AMBER_CMD_PROCESS_ROUTES = true
-          puts Amber::Server.routes.to_json
-        CODE
-        routes_json = `crystal eval #{code.gsub("\n", ";").inspect}`
-        return JSON.parse(routes_json) unless routes_json.empty?
-        puts "No routes defined!"
+      private def set_route(l)
```
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: can you provide a better argument name than simply `l` something more descriptive

File: src/amber_cmd/commands/routes.cr [hunk 12]
```diff
@@ -23,30 +35,79 @@ module Amber::CMD
...
+          route["Controller"] = md[3]?.to_s
+          route["Action"] = md[4]?.to_s
+          route["Pipeline"] = current_pipe.to_s
+          route["Scope"] = current_scope.to_s
+          route["Resource"] = md[2]?.to_s
+          routes << route
+        end
+      end
+      
+      private def set_resources(l)
```
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: Descriptive argument name

File: src/amber_cmd/commands/routes.cr [hunk 13]
```diff
@@ -23,30 +35,79 @@ module Amber::CMD
...
+          filter_actions = md[5]?.to_s.gsub(/\:|\s/, "").split(",")
+          ACTION_MAPPING.each do |verb, v|
+            v.each do |action|
+              case filter
+              when "only"
+                next unless filter_actions.includes?(action)
+              when "except"
+                next if filter_actions.includes?(action)
+              end
+              route = {"Verb" => verb}
```
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: This can be substracted to a method. See above.

File: src/amber_cmd/commands/routes.cr [hunk 14]
```diff
@@ -23,30 +35,79 @@ module Amber::CMD
...
+          route["Action"] = md[4]?.to_s
+          route["Pipeline"] = current_pipe.to_s
+          route["Scope"] = current_scope.to_s
+          route["Resource"] = md[2]?.to_s
+          routes << route
+        end
+      end
+      
+      private def set_resources(l)
+        if md = l.to_s.match(RESOURCE_ROUTE_REGEX)
```
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: Instead of wrapping it in an if statement I think it would be cleaner to turn this into a guard clause

Something like
```cr
return unless l.to_s.match(RESOURCE_ROUTE_REGEX)
```

File: src/amber_cmd/commands/routes.cr [hunk 15]
```diff
@@ -23,30 +35,79 @@ module Amber::CMD
...
+              route["Scope"] = current_scope.to_s
+              route["Resource"] = build_resource(base_route, action, current_scope)
+              routes << route
+            end
+          end
+        end
+      end
+
+      private def build_resource(route, action, scope)
+        route_end = {"show" => ":id", "new" => "new", "edit" => ":id/edit", "update" => ":id", "destroy" => ":id"}
```
Comment (eliasjpr) [inline src/amber_cmd/commands/routes.cr]: rename this to `resources_params` or `path_params`
